gio/src/unix_fd_list.rs

```diff
@@ -47,41 +47,107 @@ pub trait UnixFDListExtManual: sealed::Sealed + IsA<UnixFDList> + Sized {
     }
 
     #[doc(alias = "g_unix_fd_list_get")]
-    fn get(&self, index_: i32) -> Result<RawFd, glib::Error> {
+    fn get(&self, index_: i32) -> Result<OwnedFd, glib::Error> {
         unsafe {
             let mut error = ptr::null_mut();
-            let ret = ffi::g_unix_fd_list_get(self.as_ref().to_glib_none().0, index_, &mut error);
+            let raw_fd =
+                ffi::g_unix_fd_list_get(self.as_ref().to_glib_none().0, index_, &mut error);
+            let fd = OwnedFd::from_raw_fd(raw_fd);
             if error.is_null() {
-                Ok(ret)
+                Ok(fd)
             } else {
                 Err(from_glib_full(error))
             }
         }
     }
 
     #[doc(alias = "g_unix_fd_list_peek_fds")]
-
-    fn peek_fds(&self) -> Vec<RawFd> {
+    fn peek_fds(&self) -> &[BorrowedFd<'_>] {
         unsafe {
             let mut length = mem::MaybeUninit::uninit();
-            let ret = FromGlibContainer::from_glib_none_num(
-                ffi::g_unix_fd_list_peek_fds(self.as_ref().to_glib_none().0, length.as_mut_ptr()),
+            std::slice::from_raw_parts(
+                ffi::g_unix_fd_list_peek_fds(self.as_ref().to_glib_none().0, length.as_mut_ptr())
+                    as *const BorrowedFd,
                 length.assume_init() as usize,
-            );
-            ret
+            )
         }
     }
+
     #[doc(alias = "g_unix_fd_list_steal_fds")]
-    fn steal_fds(&self) -> Vec<RawFd> {
+    fn steal_fds(&self) -> FdArray {
         unsafe {
             let mut length = mem::MaybeUninit::uninit();
-            let ret = FromGlibContainer::from_glib_full_num(
-                ffi::g_unix_fd_list_steal_fds(self.as_ref().to_glib_none().0, length.as_mut_ptr()),
-                length.assume_init() as usize,
-            );
-            ret
+
+            let ptr =
+                ffi::g_unix_fd_list_steal_fds(self.as_ref().to_glib_none().0, length.as_mut_ptr());
+
+            FdArray {
+                ptr: ptr::NonNull::new(ptr).unwrap(),
+                len: length.assume_init() as usize,
+            }
         }
     }
 }
 
 impl<O: IsA<UnixFDList>> UnixFDListExtManual for O {}
+
+pub struct FdArray {
+    ptr: ptr::NonNull<libc::c_int>,
+    len: usize,
+}
+
+pub struct FdIterator {
+    ptr: ptr::NonNull<libc::c_int>,
+    len: usize,
+}
+
+impl Iterator for FdIterator {
+    type Item = OwnedFd;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.len > 0 {
+            let current = self.ptr.as_ptr();
+            if self.len > 1 {
+                let next = unsafe { self.ptr.as_ptr().add(1) };
+                self.ptr = ptr::NonNull::new(next).unwrap();
```

Comment:
This doesn't work. You'll have to keep track of the original pointer so it can be freed on `Drop` (of the `FdIterator`)